protocol-units/dispute/src/MCR.sol

```diff
@@ -0,0 +1,165 @@
+    }
+
+    struct Proof {
+        bytes32 blockHash;
+        bool isValid;
+        bool exists;
+    }
+
+    enum DisputeState {
+        SUBMITTED,
+        VALID,
+        INVALID,
+        ACCEPTED,
+        REJECTED,
+        CORRECT,
+        INCORRECT,
+        UNVERIFIABLE
+    }
+
+    uint256 public constant SECONDS_IN_DAY = 86400;
+    uint256 public constant SECONDS_IN_MINUTE = 60;
+
+    uint256 public currentEpoch;
+    uint256 public epochDuration;
+    uint256 public constant MIN_STAKE = 1 ether;
+    uint256 public delta = 1 * SECONDS_IN_DAY;
+    uint256 public p = 1 * SECONDS_IN_MINUTE;
+    uint256 public supermajorityStake;
+    uint256 public epochStartTimestamp;
+    uint256 public totalStake;
+
+
+    mapping(address => Validator) public validators;
+    mapping(bytes32 => Dispute) public disputes;
+    mapping(bytes32 => Proof) public verifiedProofs;
+    mapping(bytes32 => OptimisticCommitment) public optimisticCommitments;
+    mapping(uint256 => OptimisticCommitment) public epochs;
+
+    IRiscZeroVerifier public verifier;
+
+    event ValidatorRegistered(address indexed validator, uint256 stake);
+    event ValidatorDeregistered(address indexed validator);
+    event DisputeSubmitted(bytes32 indexed disputeHash, bytes32 blockHash, address indexed submitter);
+    event DisputeResolved(bytes32 indexed disputeHash, DisputeState state);
+    event ProofSubmitted(bytes32 indexed blockHash, bool isValid);
+    event ProofVerified(bytes32 indexed blockHash, bool isValid);
+    event BlockAccepted(bytes32 indexed blockHash, bytes stateCommitment);
+    event OptimisticCommitmentSubmitted(bytes32 indexed blockHash, bytes stateCommitment, uint256 validatorStake);
+
+    constructor(
+        uint256 _delta,
+        uint256 _supermajorityStake,
+        uint256 _epochDurationInDays
+    ) {
+        delta = _delta;
+        supermajorityStake = _supermajorityStake;
+        epochDuration = _epochDurationInDays * SECONDS_IN_DAY;
+    }
+
+    function updateEpoch() public {
+        uint256 epochsPassed = (block.timestamp - epochStartTimestamp) / epochDuration;
+        currentEpoch += epochsPassed;
+        epochStartTimestamp += epochsPassed * epochDuration;
+    }
+
+    function stake() external payable {
+        validators[msg.sender].isRegistered = true;
+        validators[msg.sender].stake += msg.value;
+        totalStake += msg.value;
+        emit ValidatorRegistered(msg.sender, msg.value);
+    }
+
+    function unstake(uint256 amount) external {
+        Validator storage validator = validators[msg.sender];
+        require(validator.isRegistered, "Validator not registered");
+        require(validator.stake >= amount, "Insufficient stake");
+        validator.stake -= amount;
+        totalStake -= amount;
+        if (validator.stake == 0) {
+            validator.isRegistered = false;
+        }
+        payable(msg.sender).transfer(amount);
+        emit ValidatorDeregistered(msg.sender);
```

Comment:
This event should emitted when any amount was unstaked and regardless of whether the validator has any remaining stake, correct?
So in effect, it's a confirmation that `unstake` was called by the sender with a valid argument, and it carries no other information.

Reply:
This is going to be modded progressively. For now, the above is fine, but I will be changing it fairly shortly in a manner that makes emitting an `unstake` amount more reasonable.